### PR TITLE
Fix #405: harden nhan_vien privileges and RPC boundaries

### PR DESCRIPTION
**Pull request by thienchi2109**
## Summary
- Harden direct `public.nhan_vien` table access by enabling RLS, revoking anon/authenticated table privileges, and adding explicit deny-all policies for Data API roles.
- Move `/users` management reads/writes and tenant switching writes to guarded RPC boundaries instead of direct `nhan_vien` queries.
- Add security source tests plus live SQL smoke coverage for `nhan_vien` privilege/RLS/RPC boundaries.

Fixes #405

## Scope Notes
- Supabase Dashboard admin/manual operations are not blocked by the anon/authenticated revokes; this change targets client/Data API roles and pushes app traffic through RPCs.
- `thiet_bi` is untouched.
- React Doctor reported existing/adjacent cleanup warnings in touched UI files. The oversized `UsersPage` cleanup is intentionally deferred to follow-up #425 to keep this PR narrow.
- Full forced password-change app enforcement after admin reset is tracked separately in #427. This PR sets the DB flag `password_reset_required = TRUE` during admin reset; #427 handles auth/UI enforcement.

## Live DB Changes Applied via Supabase MCP
- `20260508134826_harden_nhan_vien_privileges_and_rls`
- `20260508134945_add_nhan_vien_deny_all_rls_policies`
- `20260508141618_harden_user_create_role_and_audit`
- `20260508142115_harden_user_set_current_don_vi`
- `20260509005630_reassert_user_boundary_review_fixes`
- `20260509015738_reassert_user_boundary_review_fixes_round_2`

## Review Follow-up
- Fixed direct tenant-switch regression by hardening `user_set_current_don_vi` and routing `/api/tenants/switch` through a signed RPC call.
- Declared `/api/tenants/switch` as Node runtime because it imports `jsonwebtoken`.
- Fixed `user_create` so only admin/global callers pass the DB boundary and successful creates write `USER_CREATE` audit rows.
- Fixed `user_create` blank/whitespace password rejection before hashing, and hash generation now uses the same trimmed password value.
- Fixed non-global tenant switching so inactive tenants are rejected even when membership exists.
- Fixed admin password reset so `password_reset_required` is set to `TRUE` and the `password_reset_admin` audit insert is mandatory.
- Fixed `pg_policies.roles` smoke assertion to use set-equivalence instead of order-sensitive equality.

## Verification
- Supabase SQL smoke passed against live DB for `user_create` non-admin denial, blank password denial, trimmed-password hashing, `USER_CREATE` audit logging, tenant-switch allow/deny paths, inactive tenant denial, reset-password flagging, and reset-password audit logging; smoke transactions rolled back and left no test users.
- Live grants/RLS/policies checked for `public.nhan_vien`.
- Live function checks confirmed `user_create`, `user_set_current_don_vi`, and `reset_password_by_admin` are SECURITY DEFINER with fixed `search_path`, anon execute revoked, authenticated execute granted, and expected JWT/business guards present.
- Supabase security advisors checked after DDL. Remaining findings are baseline plus expected app-facing authenticated SECURITY DEFINER warnings for guarded RPCs.
- `git diff --check`
- `node scripts/npm-run.js run verify:no-explicit-any`
- `node scripts/npm-run.js run typecheck`
- `node scripts/npm-run.js run test:run -- src/components/__tests__/user-management.security-source.test.ts src/app/api/rpc/__tests__/rpc-whitelist.unit.test.ts src/components/__tests__/change-password-dialog.security-source.test.ts src/components/__tests__/change-password-dialog.signout.test.tsx src/auth/__tests__/auth-config.authorize-events.test.ts` → 46 tests passed.
- `node scripts/npm-run.js npx react-doctor@latest . --verbose -y --project nextn --offline --diff main` → 96/100, 19 warnings remain, deferred where out of #405 scope.

## Summary by CodeRabbit

* **New Features**
  * Admin user-management now uses centralized RPC boundaries (list, update, delete, reset) and tenant-switching via RPC.
* **Improvements**
  * Removed password field from profile editor; streamlined edit flow and safer self-action guards.
  * Hardened database boundaries and RLS for user data; stronger audit logging for admin actions.
* **Tests**
  * Added whitelist and security-source tests to enforce RPC and data-access constraints.

### CODE REVIEW

File: supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql [hunk 1]
```diff
@@ -0,0 +1,488 @@
+-- Issue #405: make public.nhan_vien an internal auth/account table.
+-- App clients must use session-signed RPC boundaries instead of direct table access.
+
+ALTER TABLE public.nhan_vien ENABLE ROW LEVEL SECURITY;
+
+REVOKE ALL ON TABLE public.nhan_vien FROM PUBLIC;
+REVOKE ALL ON TABLE public.nhan_vien FROM anon;
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql:7]: Fixed in commit 52d2116. `/api/tenants/switch` no longer updates `nhan_vien` directly with the anon client; it now signs a short-lived Supabase JWT and calls `user_set_current_don_vi`. Added a forward migration hardening that RPC with fixed `search_path`, JWT `user_id` claim matching, global active-tenant validation, and non-global membership validation. Live smoke covered allowed membership switching and cross-user denial.

File: supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    RETURN jsonb_build_object('success', false, 'message', 'Target user not found');
+  END IF;
+
+  UPDATE public.nhan_vien
+  SET hashed_password = extensions.crypt(new_password, extensions.gen_salt('bf', 12)),
+      password = 'hashed password',
+      password_changed_at = NOW(),
+      failed_attempts = 0,
+      password_reset_required = COALESCE(password_reset_required, FALSE)
+  WHERE id = p_target_user_id;
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql:333]: Valid issue, but larger than the #405 table privilege/RLS boundary. Opened follow-up #427 for forced password-change enforcement after admin reset. Keeping this PR focused on closing direct `nhan_vien` access and RPC boundary regressions.

File: supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+
+  IF p_password IS NULL OR p_full_name IS NULL OR p_role IS NULL OR p_current_don_vi IS NULL THEN
+    RAISE EXCEPTION 'Missing required fields' USING ERRCODE = '22023';
+  END IF;
+
+  IF v_app_role IS DISTINCT FROM 'global' THEN
+    IF v_req_don_vi IS NULL OR v_req_don_vi IS DISTINCT FROM p_current_don_vi THEN
+      RAISE EXCEPTION 'Không có quyền tạo người dùng ngoài đơn vị hiện tại' USING ERRCODE = '42501';
+    END IF;
+  END IF;
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql:422]: Fixed in commit 52d2116. Live DB smoke first reproduced the issue: a plain `user` claim could call `user_create` within its tenant. Added `20260508141618_harden_user_create_role_and_audit` so `user_create` now requires a JWT user id whose `nhan_vien.role` is `admin`/`global` and normalized app role `global`. Re-ran live smoke; non-admin create is denied and leaves no test rows.

File: supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    INSERT INTO public.user_don_vi_memberships(user_id, don_vi)
+    SELECT v_id, m
+    FROM unnest(p_memberships) AS m
+    WHERE m IS NOT NULL AND m <> p_current_don_vi
+    ON CONFLICT DO NOTHING;
+  END IF;
+
+  RETURN v_id;
+END;
+$$;
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql:460]: Fixed in commit 52d2116. `user_create` now inserts a `USER_CREATE` row into `public.audit_logs` with admin user, target user, target username, entity metadata, IP, user-agent, and create details. Live smoke verified the audit row inside a rolled-back test transaction.

File: supabase/tests/nhan_vien_privilege_boundaries_smoke.sql [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    'nhan_vien_deny_update',
+    'nhan_vien_deny_delete'
+  ]) AS required_policy
+  WHERE NOT EXISTS (
+    SELECT 1
+    FROM pg_policies p
+    WHERE p.schemaname = 'public'
+      AND p.tablename = 'nhan_vien'
+      AND p.policyname = required_policy
+      AND p.roles = '{anon,authenticated}'
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/tests/nhan_vien_privilege_boundaries_smoke.sql]: Fixed in commit 52d2116. The smoke test now checks `pg_policies.roles` with `@>` and `<@` set-equivalence instead of order-sensitive array equality.

File: supabase/migrations/20260508141618_harden_user_create_role_and_audit.sql [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+  IF v_admin_role IS NULL OR v_admin_role NOT IN ('admin', 'global') OR v_app_role <> 'global' THEN
+    RAISE EXCEPTION 'Access denied: Admin or Global privileges required' USING ERRCODE = '42501';
+  END IF;
+
+  IF p_username IS NULL OR v_username = '' OR POSITION(' ' IN v_username) > 0 THEN
+    RAISE EXCEPTION 'Invalid username format' USING ERRCODE = '22023';
+  END IF;
+
+  IF p_password IS NULL OR p_full_name IS NULL OR v_full_name = '' OR p_role IS NULL OR v_role = '' OR p_current_don_vi IS NULL THEN
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260508141618_harden_user_create_role_and_audit.sql:55]: Fixed in commit 43d8c27. Added forward migration `20260509005630_reassert_user_boundary_review_fixes` so `user_create` rejects NULL/blank/whitespace passwords before hashing. Live smoke covered whitespace-only password denial and left no test rows.

File: supabase/migrations/20260508142115_harden_user_set_current_don_vi.sql [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    SELECT COALESCE(dv.active, TRUE)
+    INTO v_tenant_active
+    FROM public.don_vi dv
+    WHERE dv.id = p_don_vi;
+
+    IF v_tenant_active IS DISTINCT FROM TRUE THEN
+      RAISE EXCEPTION 'Invalid or inactive tenant' USING ERRCODE = '22023';
+    END IF;
+  ELSE
+    SELECT EXISTS (
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260508142115_harden_user_set_current_don_vi.sql:50]: Fixed in commit 43d8c27. `user_set_current_don_vi` now validates target tenant active state before the global/non-global branch, so non-global users cannot switch to inactive tenants even with membership. Live smoke covered inactive tenant denial.

File: supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+  IF NOT FOUND THEN
+    RETURN jsonb_build_object('success', false, 'message', 'Target user not found');
+  END IF;
+
+  UPDATE public.nhan_vien
+  SET hashed_password = extensions.crypt(new_password, extensions.gen_salt('bf', 12)),
+      password = 'hashed password',
+      password_changed_at = NOW(),
+      failed_attempts = 0,
+      password_reset_required = COALESCE(password_reset_required, FALSE)
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260508134826_harden_nhan_vien_privileges_and_rls.sql:332]: Fixed in commit 43d8c27. `reset_password_by_admin` now sets `password_reset_required = TRUE` when resetting to the default password. Full app/auth enforcement remains tracked in follow-up #427; this PR closes the DB-side flagging gap. Live smoke verified the flag is set.

File: supabase/migrations/20260509005630_reassert_user_boundary_review_fixes.sql [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+  IF p_password IS NULL OR v_password = '' THEN
+    RAISE EXCEPTION 'Invalid password format' USING ERRCODE = '22023';
+  END IF;
+
+  IF p_full_name IS NULL OR v_full_name = '' OR p_role IS NULL OR v_role = '' OR p_current_don_vi IS NULL THEN
+    RAISE EXCEPTION 'Missing required fields' USING ERRCODE = '22023';
+  END IF;
+
+  v_hashed_password := extensions.crypt(p_password, extensions.gen_salt('bf', 12));
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260509005630_reassert_user_boundary_review_fixes.sql:65]: Fixed in commit ea17648. Superseding migration `20260509015738_reassert_user_boundary_review_fixes_round_2` now hashes `v_password` after trimming, so validation and stored bcrypt input use the same value. Live smoke verified a password submitted with leading/trailing spaces matches the trimmed value and left no smoke users behind.

[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260509005630_reassert_user_boundary_review_fixes.sql:65]: Fixed in commit ea17648. `user_create` now passes `v_password` into `extensions.crypt(...)`, and the SQL smoke covers this by creating with `  temporary-password  ` then verifying the stored hash against `temporary-password`.

File: supabase/migrations/20260509005630_reassert_user_boundary_review_fixes.sql [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+      COALESCE(pg_catalog.inet_client_addr(), '0.0.0.0'::INET),
+      COALESCE(
+        NULLIF((COALESCE(NULLIF(current_setting('request.headers', true), ''), '{}')::jsonb)->>'user-agent', ''),
+        'unknown'
+      ),
+      'nhan_vien',
+      p_target_user_id,
+      target_rec.username
+    );
+  EXCEPTION WHEN others THEN
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260509005630_reassert_user_boundary_review_fixes.sql:299]: Fixed in commit ea17648. `reset_password_by_admin` no longer wraps the `password_reset_admin` audit insert in a swallowed exception block; the audit insert is mandatory in the same function call, so an audit failure aborts the reset. Live smoke verifies the reset flag and audit row together.

File: supabase/migrations/20260509015738_reassert_user_boundary_review_fixes_round_2.sql [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+SECURITY DEFINER
+SET search_path = public, extensions, pg_temp
+AS $$
+DECLARE
+  v_claims JSONB := COALESCE(current_setting('request.jwt.claims', true), '{}')::jsonb;
+  v_role TEXT := LOWER(NULLIF(COALESCE(v_claims->>'app_role', v_claims->>'role'), ''));
+  v_claim_user_id BIGINT := NULLIF(v_claims->>'user_id', '')::BIGINT;
+  admin_rec public.nhan_vien%ROWTYPE;
+  target_rec public.nhan_vien%ROWTYPE;
+  new_password TEXT := 'userqltb';
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/migrations/20260509015738_reassert_user_boundary_review_fixes_round_2.sql:152]: Leaving this unchanged in #426 by scope decision. The fixed reset password is the current deployed hospital workflow: admin resets, communicates the temporary value through the existing operational channel, and the user changes it afterward. Generating a random DB-side password would require a broader contract/UI/operations change so the admin can receive and safely hand off that value; otherwise the reset flow would become unusable. This PR already reduces the boundary risk by setting `password_reset_required = TRUE` and auditing the reset. Full app enforcement for forced password change is tracked in #427, where the temporary-password workflow can be redesigned intentionally if needed.

File: supabase/tests/nhan_vien_privilege_boundaries_smoke.sql [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+    v_created_id := public.user_create(
+      'issue405_trimmed_password_user_smoke',
+      '  temporary-password  ',
+      'Issue 405 Trimmed Password User Smoke',
+      'user',
+      v_tenant_id,
+      NULL
+    );
+
+    SELECT nv.hashed_password
```
[the remark this answers is not in the document]
Comment (thienchi2109) [reply supabase/tests/nhan_vien_privilege_boundaries_smoke.sql:249]: Fixed in commit 149ec96. The smoke test now checks `NOT FOUND OR v_hash IS NULL` after the lookup, so it fails clearly if `user_create` does not create the expected row with a hashed password. Reran the focused live smoke against Supabase; it passed and left `leaked_smoke_users = 0`.
